### DISCUSSION TITLE
Setting Machine Learning URL in environment variables

### DISCUSSION THREAD

**Discussion by armond-avanes** [immich-app/immich#14739] [Q&A]
Is there any way to set the URL of machine learning in the docker environment variables? I'm trying to automate the initial setup/config as much as possible, but couldn't find any such setting here: https://immich.app/docs/install/environment-variables/

Any help is really appreciated. Thanks!

UPDATE: The reason why I need to override the default is that I'm running the machine learning container in another machine.

**Comment by mertalev** [comment]
You can use a config file. If you go to the settings page, you can export your current settings as a JSON file. You can mount this file into the server container and set the path within the container through the `IMMICH_CONFIG_FILE` environmental variable.

**Comment by armond-avanes** [comment]
I appreciate your response. I quite understand your solution and it should work. But ML URL is just a small part of the whole config. In my experience, this can easily break the future installations if the format is changed or new default settings are added or existing default settings are removed/changed, etc. Having this URL as an environment variable will provide much more flexible solution.  It's like database or redis host/port env variables. Just my personal preference/opinion.

**Comment by aarontc** [comment]
+1 from me, allowing this to be an environment variable will support dynamic deployments (Nomad, K8s, etc.). The support is already present to specify the database and Redis hosts by environment variables, which makes it look like an oversight to be unable to configure the machine learning host this way.

**Comment by bo0tzz** [comment]
There is in fact an env var for this, called `IMMICH_MACHINE_LEARNING_URL`, but it's deprecated and thus we don't recommend using it.

**Reply by alex3025** [reply]
Are there any plans to replace it with another solution/variable?
I also need it because I've renamed the ML container.

**Reply by bo0tzz** [reply]
Do you actually need an env var? You can set the ML URL in the admin settings.

**Reply by alex3025** [reply]
Oh, didn't see that. Exactly what I needed, thanks!

**Reply by PipeItToDevNull** [reply]
> Do you actually need an env var? You can set the ML URL in the admin settings.

Yes, my environment is dynamically deployed with Nomad and the IP of the machine is not known. I need to set a variable in the job

**Reply by bo0tzz** [reply]
Surely you could use DNS instead and configure the (fixed) DNS name in the settings.

**Reply by aarontc** [reply]
DNS can be used (at least in my Nomad setup). I think the part that puzzles some of us is why environment variables are preserved/recommended for certain dependencies but not others.

**Reply by aarontc** [reply]
Also, using DNS doesn't support dynamic port allocation unless Immich performs SRV lookups, which I do not believe is the case. This makes Nomad or other dynamic deployments slightly more complex.
